plugins/pluginmanagers/lz-n.nix

```diff
@@ -0,0 +1,106 @@
+{
+  lib,
+  helpers,
+  config,
+  pkgs,
+  ...
+}:
+with lib;
+let
+  lzPluginType =
+    with helpers.nixvimTypes;
+    submodule {
+      options = {
+        name = mkOption {
+          type = str;
+          description = "The plugin's name (not the module name). This is what is passed to the load(name) function.";
+        };
+        enabledInSpec = helpers.mkNullOrStrLuaFnOr bool ''
+          When false, or if the function returns false, then this plugin will not be included in the spec.
+          This option corresponds to the `enabled` property of lz.n.
+        '';
+        beforeAll = helpers.mkNullOrLuaFn "Always executed before any plugins are loaded.";
+        before = helpers.mkNullOrLuaFn "Executed before this plugin is loaded.";
+        after = helpers.mkNullOrLuaFn "Executed after this plugin is loaded.";
+        event = helpers.mkNullOrOption (either str (listOf str)) "Lazy-load on event. Events can be specified as BufEnter or with a pattern like BufEnter *.lua";
+        cmd = helpers.mkNullOrOption (either str (listOf str)) "Lazy-load on command.";
+        ft = helpers.mkNullOrOption (either str (listOf str)) "Lazy-load on filetype.";
+        colorscheme = helpers.mkNullOrOption (either str (listOf str)) "Lazy-load on colorscheme.";
+        keys = helpers.mkNullOrOption (listOf helpers.keymaps.mapOptionSubmodule) "Lazy-load on key mapping. Use the same format as `config.keymaps`.";
+        priority = helpers.mkNullOrOption number ''
+          Only useful for start plugins (not lazy-loaded) to force loading certain plugins first. 
+          Default priority is 50 (or 1000 if colorscheme is set).
+        '';
+        load = helpers.mkNullOrLuaFn "Can be used to override the `vim.g.lz_n.load()` function for this plugin.";
+      };
+    };
+in
+helpers.neovim-plugin.mkNeovimPlugin config {
+  name = "lz-n";
+  originalName = "lz.n";
+  maintainers = [ helpers.maintainers.psfloyd ];
+  defaultPackage = pkgs.vimPlugins.lz-n;
+
+  settingsDescription = ''
+    Options provided to `vim.g.lz_n`.
+
+    `{ load = "fun"; }` -> `vim.g.lz_n = { load = fun, }`
+  '';
+
+  settingsOptions = {
+    load = helpers.defaultNullOpts.mkLuaFn "vim.cmd.packadd" ''
+      Function used by `lz.n` to load plugins.
+    '';
+  };
+
+  callSetup = false; # Does not use setup
+
+  extraOptions = {
+    plugins = mkOption {
+      description = "List of plugins processed by lz.n";
+      default = [ ];
+      type = types.listOf lzPluginType;
```

Comment:
It'd be good if this accepted "rawLua" types, e.g. using `helpers.nixvimTypes.maybeRaw` or by adding `helpers.nixvimTypes.rawLua` to a `lib.types.oneOf`.

Reply:
I think it'd also be fine to have `coercedTo str toList (listOf str)` here, since:

a) upstream accept either str or table of str
b) we need a list type for merging

Reply:
> It'd be good if this accepted "rawLua" types, e.g. using `helpers.nixvimTypes.maybeRaw` or by adding `helpers.nixvimTypes.rawLua` to a `lib.types.oneOf`.

The problem I found with using `maybeRaw` o `oneOf` is that documentation does not get generated for `lz-n.plugins.*.option`.

This is what I tested:

```nix
type = with helpers.nixvimTypes; maybeRaw (coercedTo str toList (listOf lzPluginType));
```

```nix
type = with helpers.nixvimTypes; oneOf [ rawLua (coercedTo str toList (listOf lzPluginType))];
```

Reply:
#1618 is the issue in question. We're struggling to get to the bottom of it, but if a type really should be a maybeRaw submodule, we should make it such.

One caveat: if a type is `attrs` or `attrsOf anything` then `{__raw = "";}` will pass the `check` function. While it'd be undocumented, it would be supported to use raw lua with such a type even without wrapping it in `maybeRaw`.

Another workaround is to use a custom type with a custom `check` function, but this is cumbersome.